exec/session.go

```diff
@@ -272,6 +272,8 @@ var statusMu sync.Mutex
 
 func (s *Session) run(ctx context.Context, calldepth int, funcv *bigslice.FuncValue, args ...interface{}) (*Result, error) {
 	location := "<unknown>"
+	runContext, runContextCancel := context.WithCancel(ctx)
+	defer runContextCancel()
```

Comment:
It seems like this will cause machine management to stop when this cancellation is executed when returning from this method, as this context is now the context passed to `(*sliceMachine).Go`. This is undesirable, as `Session`s can be used to run multiple invocations.

I think we want to continue to log the `lost machine...` message as long as the session has not been shut down, as task results can be reused by future invocations, and it's reasonably informative to know that they will need to be recomputed. I think we should look to only clean up the messages that folks see after session shutdown, as lost tasks have no observable effect at that point.

Reply:
Alright. This brings up another issue however. Do you think it is alright to store a context on the session struct (I earlier found "Do not store Contexts inside a struct type" in the documentation). If not do you think there is another way to do this? The only other way I could think of is to require the user to create their own context and pass it to each session method.

Reply:
One possibility is to keep a channel that we close on shutdown.

Reply:
I've just put up a channel based version of this. Let me know what you think.

Reply:
I think the problem is unresolved. My reading is as follows:
1. `(*Session).run` now creates a `Context`, which I'll call `ctx` henceforth. It is cancelled when`(*Session).run` returns.
2. `(*Session).run` passes `ctx` to `Eval`.
3. `Eval` passes `ctx` to `executor.Run`. Suppose `executor` is a `*bigmachineExecutor`.
4. `(*bigmachineExecutor).Run` passes `ctx` to `(*bigmachineExecutor).manager`.
5. `(*bigmachineExecutor).manager` passes `ctx` to `(*machineManager).Do`.
6. `(*machineManager).Do` passes `ctx` to `startMachines`.
7. `startMachines` passes `ctx` to `(*sliceMachine).Go`.

When `(*Session).run` returns, both:
* `(*machineManager).Do` will return, thereby no longer managing machines.
* `(*sliceMachine).Go` will return (for all machines).

Neither of these things should happen, as the `*Session` needs to continue to be usable.

Reply:
As of the current diff `(*slicemachine).Go` returns based on a channel which is closed on session shutdown. Pushing up a diff now which will make `(*machineManager).Do` exit based on this channel as well.

Reply:
I've leaved the context plumbed through but it won't deactivate those processes when its closed.

Reply:
When the context is cancelled, the channel returned by `ctx.Done()` will be closed, so the `select` will unblock. When we check the `for`-loop condition, `ctx.Err()` will be non-`nil` and `(*sliceMachine).Go` will return. Could you help me understand how this reasoning is invalid?

Reply:
Ah I forgot about the for loop condition I was just looking at the select at the bottom. I just updated it to use background context and double checked on adhoc.